private/buf/bufsync/bufsyncapi/sync_handler.go

```diff
@@ -0,0 +1,526 @@
+					RepositoryId: repositoryID,
+					Name:         tag,
+					CommitName:   commitRes.Msg.Reference.GetVcsCommit().CommitName,
+				}))
+				if err != nil {
+					return fmt.Errorf("create new tag %q on module %q: %w", tag, moduleTags.TargetModuleIdentity().IdentityString(), err)
+				}
+			} else {
+				// TODO: don't do this unless we need to
+				_, err := repositoryTagService.UpdateRepositoryTag(ctx, connect.NewRequest(&registryv1alpha1.UpdateRepositoryTagRequest{
+					RepositoryId: repositoryID,
+					Name:         tag,
+					CommitName:   &commitRes.Msg.Reference.GetVcsCommit().CommitName,
+				}))
+				if err != nil {
+					return fmt.Errorf("update existing tag %q on module %q: %w", tag, moduleTags.TargetModuleIdentity().IdentityString(), err)
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (h *syncHandler) SyncModuleBranch(ctx context.Context, moduleBranch bufsync.ModuleBranch) error {
+	for _, moduleCommit := range moduleBranch.CommitsToSync() {
+		bucket, err := moduleCommit.Bucket(ctx)
+		if err != nil {
+			return fmt.Errorf("read bucket for commit %q: %w", moduleCommit.Commit().Hash(), err)
+		}
+		syncPoint, err := h.pushOrCreate(
+			ctx,
+			moduleCommit.Commit(),
+			moduleBranch.BranchName(),
+			moduleCommit.Tags(),
+			moduleBranch.TargetModuleIdentity(),
+			bucket,
+		)
+		if err != nil {
+			// We failed to push. We fail hard on this because the error may be recoverable
+			// (i.e., the BSR may be down) and we should re-attempt this commit.
+			return fmt.Errorf(
+				"failed to push or create %s at %s: %w",
+				moduleBranch.TargetModuleIdentity().IdentityString(),
+				moduleCommit.Commit().Hash(),
+				err,
+			)
+		}
+		_, err = h.container.Stderr().Write([]byte(
+			// from local                                        -> to remote
+			// <module-directory>:<git-branch>:<git-commit-hash> -> <module-identity>:<bsr-commit-name>
+			fmt.Sprintf(
+				"%s:%s:%s -> %s:%s\n",
+				moduleBranch.Directory(), moduleBranch.BranchName(), moduleCommit.Commit().Hash().Hex(),
+				moduleBranch.TargetModuleIdentity().IdentityString(), syncPoint.BsrCommitName,
+			)),
+		)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (h *syncHandler) IsProtectedBranch(
+	ctx context.Context,
+	moduleIdentity bufmoduleref.ModuleIdentity,
+	branchName string,
+) (bool, error) {
+	// If the branch is the Git default branch, protect it.
+	if branchName == h.repo.DefaultBranch() {
+		return true, nil
+	}
+	return h.IsReleaseBranch(ctx, moduleIdentity, branchName)
+}
+
+func (h *syncHandler) IsReleaseBranch(
+	ctx context.Context,
+	moduleIdentity bufmoduleref.ModuleIdentity,
+	branchName string,
+) (bool, error) {
+	// Otherwise the only other protected branch is the Repository's default (release) branch.
+	// We cache a repository's default branch even though it can change because it's _extremely_ unlikely that it changes.
```

Comment:
Outdated comment. This func doesn't refer to anything "protected", just the release branch.